subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/SelectorState.java

```diff
@@ -149,12 +145,24 @@ public void select(ComponentState selected) {
         this.selected = selected;
     }
 
-    private ComponentSelectionDescriptorInternal selectionDescriptionForDependency(DependencyMetadata dependencyMetadata) {
-        ComponentSelectionDescriptorInternal selectionDescriptor = dependencyMetadata.isPending() ? CONSTRAINT : REQUESTED;
-        if (dependencyMetadata.getReason() != null) {
-            selectionDescriptor = selectionDescriptor.withReason(dependencyMetadata.getReason());
+    /**
+     * Overrides the component that is the chosen for this selector.
+     * This happens when the `ModuleResolveState` is restarted, during conflict resolution or version range merging.
+     */
+    public void overrideSelection(ComponentState selected) {
+        if (this.selected == null) {
+            // Do not override if this selector hasn't yet been resolved
+            return;
```

Comment:
Sounds strange that we ever call `overrideSelection` if `selected` != `null`, but that's independent from this refactor.

Reply:
Yep, I want to keep cleaning this up. In fact `SelectorState.selected` may no longer be required, since we choose a single selected component for the _module_, and all selectors belong to a module. So the only states allowed for a selector will be:
-  'failed': with an associated failure message
- 'succeeded': which means it helped to select the component for module
- 'not-yet-evaluated': which means this is a newly discovered selector and we need to re-resolve the set of selectors for the module